core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -161,7 +240,25 @@ private JoinNode getSizeBasedJoin(JoinNode joinNode, Context context)
 
         if (isLeftSideSmall) {
             // left side is small enough, but must be partitioned
-            return joinNode.flipChildren().withDistributionType(PARTITIONED);
+            return flippedJoin.withDistributionType(PARTITIONED);
+        }
+
+        // Flip join sides if one side is smaller than the other by more than SIZE_DIFFERENCE_THRESHOLD times.
+        // We use 8x factor because the reduction in size from aggregation and filtering
+        // may not be fully accounted here.
+        double leftSourceSize = getApproximateSourceSizeInBytes(joinNode.getLeft(), context);
+        double rightSourceSize = getApproximateSourceSizeInBytes(joinNode.getRight(), context);
+        if (rightSourceSize * SIZE_DIFFERENCE_THRESHOLD < leftSourceSize) {
+            if (mustReplicate(joinNode, context)) {
```

Comment:
Potentially we wouldn't have to replicate for flipped join (left vs right join). However, replication was potentially rejected in code above due to excessive size.

Maybe only consider partitioned cases (assuming that replicated cases were rejected above)?

Reply:
> Potentially we wouldn't have to replicate for flipped join (left vs right join)

If the replication is due to right side being a scalar, then that's a good choice and this case was likely already captured in the previous cases about size < 100MB.
If the replication is due to cross-join then it'll be the same even if the join is flipped.

Reply:
I've simplified this is a bit more to remove the non-flipped case as that's not really needed.